### DISCUSSION TITLE
passing in parameterse to `otp_qr_code` method?

### DISCUSSION THREAD

**Discussion by benkoshy** [jeremyevans/rodauth#256] [General]
Hi there,

Is it a crazy idea to allow for the passing in of parameters to a delegated method:

https://github.com/jeremyevans/rodauth/blob/master/lib/rodauth/features/otp.rb#L310-L313

```rb
    def otp_qr_code(color:  "ffffff")
      svg = RQRCode::QRCode.new(otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true, color: color)
      svg.sub(/\A<\?xml version="1\.0" standalone="yes"\?>/, '')
    end
```

### Context: Dark Mode
I wish  to change the fill / colors of the QR code, for users who are using Dark Mode.

![image](https://user-images.githubusercontent.com/15097447/186039739-db91b4e7-2c3c-4257-802d-07facad20971.png)

### Current Solution

Is to just re-implement the method in the view, and pass in what I need to.

```erb
<%== RQRCode::QRCode.new(rodauth.otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true, color: "ffffff") %>
```
![image](https://user-images.githubusercontent.com/15097447/186039955-7d8329b9-d2ee-49d7-a1e5-6939792d86a1.png)

**Comment by jeremyevans** [comment]
Adding arguments to the method (even optional arguments) cannot done without breaking backwards compatibility.  We could consider adding a method to specify the color to use, so you only need to override that method to change the color.

**Reply by benkoshy** [reply]
Rather than adding another method, would it be impractical to make explicit what is being done implicitly: to set the fill / color to their respective defaults (i.e. black / white -- I forget the order).  That will solve my particular problem, while not breaking backward compatibility? good / bad idea?

If that is acceptable, I can provide the PR.

**Reply by jeremyevans** [reply]
I would prefer not to provide explicit defaults that match the rqrcode current implicit defaults, unless there is a way to  configure the values in Rodauth. I'm not sure how specifying explicit defaults would solve your problem.  However, maybe I'm not understanding what you are asking for.  Can you provide a short diff?

**Reply by benkoshy** [reply]
>  Can you provide a short diff?

The diff will be implicit in below:

If no explicit default values are provided:

```erb
<div class="dark">
       <p><%== rodauth.otp_qr_code %></p>
</div>
```
Then this is the result (Cameras will have a hard time scanning this code):

![image](https://user-images.githubusercontent.com/15097447/186095657-2edcc228-82e4-4d4f-a8ab-fa99c915aa92.png)

However, if I pass in some sensible defaults, this is the result:

```erb
<div class="dark">
   <%== RQRCode::QRCode.new(rodauth.otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true, color: "ffffff", fill: "#000000") %> 
</div>
```

Cameras would have a much easier time recognising the above.

![image](https://user-images.githubusercontent.com/15097447/186095722-c242d792-f31b-460f-9f1c-08ead67a8f8b.png)

The problem is essentially "solved" if we explicitly specify a white / black background and foreground color (which was what we originally had) + this has the the added advantage of it just 'working' in "dark" mode.

Please let me know if you need further info / or a working example, or otherwise etc.

**Reply by jeremyevans** [reply]
So in the form of a short diff, what you are requesting is?:

```diff
diff --git a/lib/rodauth/features/otp.rb b/lib/rodauth/features/otp.rb
index 4aad932..de36bd1 100644
--- a/lib/rodauth/features/otp.rb
+++ b/lib/rodauth/features/otp.rb
@@ -308,7 +308,7 @@ module Rodauth
     end
 
     def otp_qr_code
-      svg = RQRCode::QRCode.new(otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true)
+      svg = RQRCode::QRCode.new(otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true, :color => "fff", :fill => "#000000")
       svg.sub(/\A<\?xml version="1\.0" standalone="yes"\?>/, '')
     end
 
```

Basically, rqrcode uses a black foreground and transparent background by default, and that causes a problem in dark mode where the background is dark.

Personally, I don't like the color being flipped, but I would be OK specifying `fill: '#fff'`, so that the background is white instead of transparent.  Would this work for you?:


```diff
diff --git a/lib/rodauth/features/otp.rb b/lib/rodauth/features/otp.rb
index 4aad932..21c4f2c 100644
--- a/lib/rodauth/features/otp.rb
+++ b/lib/rodauth/features/otp.rb
@@ -308,7 +308,7 @@ module Rodauth
     end
 
     def otp_qr_code
-      svg = RQRCode::QRCode.new(otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true)
+      svg = RQRCode::QRCode.new(otp_provisioning_uri).as_svg(:module_size=>8, :viewbox=>true, :use_path=>true, fill: "#fff")
       svg.sub(/\A<\?xml version="1\.0" standalone="yes"\?>/, '')
     end
 
```

**Reply by benkoshy** [reply]
> Personally, I don't like the color being flipped, but I would be OK specifying fill: '#fff', so that the background is white instead of transparent. Would this work for you?:

yes you got it perfectly. I forget which key it was (whether "fill" or "color) - but that aside, we have a perfect meeting of the minds.

I raise this issue mainly for the benefit of the community and leave it to your absolute discretion whether the proposal has merit or should be discarded.

**Reply by jeremyevans** [reply]
Adding the `fill` option seems fine, I'll push a commit for it shortly.

**Reply by benkoshy** [reply]
>  ... I'll push a commit for it shortly.

I do not propose to increase burdens on maintainers, I could made a PR myself.

Thank you for your work on this (and other) gems.
